torch/export/dynamic_shapes.py

```diff
@@ -904,6 +909,149 @@ def check_shape(path, t, dynamic_shape):
+                "Dim/DerivedDims expect all equal or related dimensions to be specified, and does not yet compose well with `True`. " \
+                "We suggest using `True` mixed with `None` for auto-dynamic + static shapes, plus torch._check(dim >= min), "
+                "torch._check(dim <= max) calls in your program to specify min/max ranges, or `Dim`/`DerivedDim` mixed with `None` " \
+                "if you want to assert on the exact specification of your program's dynamic shapes behavior.",
+                case_name="dynamic_shapes_validation",
+            )
+
+
+def _transform_shapes_for_default_dynamic(
+    combined_args: Dict[str, Any],
+    dynamic_shapes: Union[Dict[str, Any], Tuple[Any], List[Any], None],
+) -> Union[Dict[str, Any], Tuple[Any], List[Any], None]:
+    """
+    In the long run this might not be needed, but this exists because export.export() and _dynamo.export()
+    historically have different semantics for how dynamic_shapes are specified, but go through the same
+    process of producing constraints, and now both use assume_static_by_default=False.
+
+    For _dynamo.export(), the semantics for dynamic_shapes are:
+    - None: dynamic, allocated a symbol
+    - Dim/DerivedDim: a strict assertion on the min/max range for this symbol, and require a specification
+      for all dims governed by this symbol (i.e. relations, equality, linear relations, etc.)
+
+    For export.export(), historically dynamism for unspecified dims has been undesirable, so the semantics are:
+    - True: dynamic, allocated a symbol
+    - None/unspecified: static
+    - Dim/DerivedDims: also a strict assertion
+
+    To allow both APIs to follow the same process for producing constraints, this function converts dynamic_shapes
+    for export.export() to be compatible with _process_dynamic_shapes() and assume_static_by_default=False, turning them
+    into essentially what they'd look like for _dynamo.export().
+
+    An example conversion might look like, for a 3-d input tensor:
+
+        input spec: {
+            0: True,
+            1: None,
+            2: Dim("dx"),
+        }
+        output spec: {
+            0: None,  # None: dynamic by default
+            1: 32,  # explicitly provide static shape
+            2: Dim("dx"),  # remains the same
+        }
+    """
+
+    def _tree_map_helper(tree, val):
+        """
+        If the user generally specifies dynamic_shapes=None for a pytree input,
+        we'd like to convert this into a tree of Nones following the input spec,
+        so we can explicitly specify static dims for all tensor dimensions.
+        Non-builtin types for pytree (e.g. custom dataclasses) creates some difficulty,
+        in which case the correct format is a list containing specs for each child attribute.
+        """
+        if (node_type := _get_node_type(tree)) not in SUPPORTED_NODES:  # is_leaf
```

Comment:
There's some similar logic in `_tree_map_with_path`, have you cross-checked? In particular, for registered pytree nodes we match whatever `flatten` gives. Since you mentioned a failing test that involves dataclasses, this could be an area where there's maybe something extra or missing wrapping?

Reply:
Yeah, I guess this is basically _tree_map_with_path + we want to unflatten as a list when the type isn't builtin - for some reason that's the correct spec for dynamic_shapes